### ISSUE TITLE
Audio loss on gts4lv with TWRP

### ISSUE THREAD

**Issue by Deewiant** [topjohnwu/Magisk#5860]
Audio doesn't work at all when using a TWRP image patched by Magisk.

Device: Samsung Galaxy Tab S5e LTE (SM-T725) (gts4lv)
Android version: 12 (LineageOS 19.1, 20220521-nightly)
Magisk version name: 8dfb30fe
Magisk version code: 24313

Original TWRP image: twrp-3.6.1_9-0-gts4lv.img from https://dl.twrp.me/gts4lv/

adb logcat from a boot before installing Magisk: [without-magisk.log](https://github.com/topjohnwu/Magisk/files/8781208/without-magisk.log)
Magisk log from patching the image: [magisk_log_2022-05-26T21_05_45.log](https://github.com/topjohnwu/Magisk/files/8781206/magisk_log_2022-05-26T21_05_45.log)
adb logcat from a boot after flashing the patched recovery: [with-magisk.log](https://github.com/topjohnwu/Magisk/files/8781207/with-magisk.log)
Magisk log from that same boot: [magisk_log_2022-05-26T21_10_37.log](https://github.com/topjohnwu/Magisk/files/8781205/magisk_log_2022-05-26T21_10_37.log)

(logcats were stopped right after booting, unlocking, and tapping notification sounds on/off to confirm whether audio works.)

Looking at the logcats, the immediate problem seems to be here:

```
D audio_hw_utils: audio_extn_utils_open_snd_mixer: retry, retry_num 1
[..]
D audio_hw_utils: audio_extn_utils_open_snd_mixer: retry, retry_num 100
E audio_hw_utils: audio_extn_utils_open_snd_mixer: Unable to find correct sound card, aborting.
```

Which then culminates in this constant spam while the system is up:

```
W AudioManager: updateAudioPortCache: listAudioPorts failed
E AudioSystem-JNI: AudioSystem::listAudioPorts error -19
```

Without Magisk none of that happens, instead it finds the card:

```
D audio_hw_utils: audio_extn_utils_open_snd_mixer: snd_card_name: sdm670-snd-card
```

I only use Magisk for root, and that seems to work fine, along with the rest of the system apart from audio.

(Apparently this was previously reported in #5389 but that got auto-closed, hopefully the bot likes this one better.)

**Comment by Deewiant** [comment]
As usual the solution appeared only after writing about the problem: I installed the Magisk APK (the same debug 24313 version) with `adb sideload` in TWRP i.e. using the deprecated "Custom Recovery" install method, and that seems to have done the trick for whatever reason. I now have both Magisk and audio working fine.

**Comment by pndwal** [comment]
But did you follow official instructions from Dev maintaining official TWRP for Samsung Galaxy Tab S5e LTE from Team Win's site for installing a Magisk-patched version their custom recovery?...

Seems that what you have effectively done is to reinstall Magisk after first rooting your device using the requisite patch / flash method...

Installing APK as zip installer via ADB sideload _would not otherwise work_ as it cannot patch ramdisk in recovery partition; that was already done.

Overflashing Magisk seems to have corrected issues resulting from installing Magisk-patched TWRP before an initial installation of Magisk (in stock recovery) on your A only legacy SAR device where both TWRP and Magisk-rooted Android boot using recovery ramdisk...

The proper method is simply to root prior to flashing Magisk-patched custom recovery and avoid the need to fix issues. Both the official method and reasons are covered here:
> You are therefore strongly advised to first [root your device with Magisk](https://topjohnwu.github.io/Magisk/install.html#samsung-system-as-root) and then pre-root a TWRP image before flashing it to recovery. The usual path of installing TWRP prior to rooting is problematic on this device, because the Magisk installer does not support installation to recovery from TWRP. Furthermore, prior rooting of the device with Magisk carries the significant advantage of disabling various device protection mechanisms that you will otherwise need to take care of manually...

https://twrp.me/samsung/samsunggalaxytabs5elte.html

**Comment by Deewiant** [comment]
I had previously been running LineageOS 18.1, and on first installing it I did go through those [Samsung SAR install instructions](https://topjohnwu.github.io/Magisk/install.html#samsung-system-as-root). Now I initially only dirty-flashed 19.1 on top of it, per the [LineageOS upgrade instructions](https://wiki.lineageos.org/devices/gts4lv/upgrade), which is when I originally ran into these issues.

After that, I figured I'd do a clean flash of everything, at which I point I did re-read those instructions for the first time in a long while and quickly got confused. My impression was:

- > Unlocking Bootloader
  
  Already done, so I don't need to worry about it.

- > Choose “Select and Patch a File” in method, and select the AP tar file
  
  I am not running the Samsung stock firmware and don't want to be either, so I should be patching TWRP instead, and I can flash it directly with my previous TWRP instead of Odin. Doing this matches up also with the "pre-root a TWRP image before flashing it to recovery" line that you quoted from https://twrp.me/samsung/samsunggalaxytabs5elte.html since I did in fact have a rooted device to begin with.

Maybe I misstepped there and should have done a full factory reset back to the Samsung stock firmware and then followed all the steps from the very beginning?

I ended up actually doing so at one point in my adventures, using Frija to get the latest build, but at that point I think I had forgotten about most of those pages and instead followed the [LineageOS install instructions](https://wiki.lineageos.org/devices/gts4lv/install) to first install the LineageOS recovery and then used Odin to flash TWRP on top of it (at this point unpatched by Magisk, I think).

I admit I don't really understand enough of these low-level Android internals. It's very unclear to me what whole boot/recovery setup exactly is, what operations affect what parts of it and how, and the difference between having root in the booted system vs. having "rooted the device" in any number of ways. If I should have re-installed a Magisk-patched _stock_ recovery and done a full factory reset when upgrading LineageOS then fair enough, but that wasn't at all obvious to me.

**Comment by pndwal** [comment]
Well you didn't have root when flashing patched-TWRP after a dirty LOS flash if it installed LOS recovery... if you are able to de-select recovery from the flash and did so you should have retained TWRP w/ root, but I'm guessing flash included LOS recovery...

Flashing back stock can only be done w/ full Odin restore with data wipe, but doubt it's necessary. Assuming recovery was overwritten, your procedure should have been root LOS as you would normally (I'm guessing LOS binaries for Samsung w/ SAR are in AP, BL, and CP as w/ stock, and that you patch AP and Odin flash all these together with HOME_CSC per John's instructions?... I don't know), then, with root, patch and flash TWRP...

Nb. addon.d survival script can only work on devices w/ boot ramdisk, so as a general rule Magisk won't survive an OTA update.

**Comment by Deewiant** [comment]
> Well you didn't have root when flashing patched-TWRP after a dirty LOS flash if it installed LOS recovery... if you are able to de-select recovery from the flash and did so you should have retained TWRP w/ root, but I'm guessing flash included LOS recovery...

The initial dirty flash I did preserved TWRP, it does not include the LOS recovery. I'm pretty sure I did retain root as well, but I quickly noticed the audio issue and I didn't think it was Magisk-related at the time so I didn't really pay attention to root.

> Flashing back stock can only be done w/ full Odin restore with data wipe, but doubt it's necessary.

This is what I did with that Frija download I mentioned.

> Assuming recovery was overwritten, your procedure should have been root LOS as you would normally (I'm guessing LOS binaries for Samsung w/ SAR are in AP, BL, and CP as w/ stock, and that you patch AP and Odin flash all these together with HOME_CSC per John's instructions?... I don't know), then, with root, patch and flash TWRP...

The way the LOS instructions explain it, you flash a custom recovery (of course the LOS recovery is assumed) into AP with Odin, and then use that recovery to install LOS itself. BL/CP/CSC are only used with stock.

The Magisk instructions for Samsung SAR devices seem to assume the stock ROM, so this is where it gets confusing for me. From an already rooted LOS + TWRP, should I have:

1. Installed stock, rooted stock, installed LOS, flashed TWRP, patched TWRP?
2. Installed LOS, flashed TWRP, patched TWRP?
3. Flashed TWRP, installed LOS, patched TWRP?
4. Flashed TWRP, patched TWRP, installed LOS?
5. Something else?

And is there a difference between using Odin vs. TWRP itself to install a patched TWRP?

The end state seems to be the same to me in any case: patched TWRP and LOS. So I don't understand how these actually differ, if they do.

At least in terms of the audio issue: I initially did the dirty flash which exhibited the issue, then I did option 3 (updating TWRP before the clean install) which exhibited the issue, then I did option 1 minus rooting stock which also exhibited the issue. From there I eventually did option 2 to get the logcats I posted, i.e. another clean install starting with an unpatched TWRP in which audio worked, and a patched TWRP which exhibited the issue again. And after that I did the sideload, which somehow fixed everything.

> Nb. addon.d survival script can only work on devices w/ boot ramdisk, so as a general rule Magisk won't survive an OTA update.

I noticed this point in the Magisk docs, but for whatever reason the LineageOS OTA updates haven't caused issues for me. Perhaps they simply don't flash recovery at least by default, so this might be another thing that applies mainly to stock. (Or then LOS is the outlier here. I haven't touched other ROMs much.)

**Comment by pndwal** [comment]
Ah, so if your LOS doesn't include recovery seems your initial method may be correct after all... I saw a few red herrings... Very sorry if I've needlessly clouded this issue. 😬

Stock OTA updates require restoring all images to stock (unpatched) first as a requirement for delta (incremental bit-matched) updates to work, and root is necessarily lost. LOS OTAs are different as are complete image updates (not incremental) and don't require Restore Images to work, but still generally remove root as they normally include the boot image (recovery.img is also a boot image) patched by Magisk. However, your case is special if ROM lacks recovery.img where Magisk is flashed...

The issue still seems to be unrelated to Magisk however, and you do need to understand a bit about those inner workings to appreciate this. I'll attempt some explanation of complex mechanisms at play... 😜

First, understand that both John Wu (Magisk) and LineageOS team have implemented methods to bypass the need for TWRP for good reasons; Dee's Troy's team (TWRP) often lags way behind, not matching Magisk's pace in adapting to Google's boot and filesystem changes, and also introduces another set of difficulties, bugs, incompatibilities, user issues etc, including possibility of users installing OS or Magisk with an incompatible version as you are. No TWRP release is Android 12 compatible yet. (Your twrp-3.6.1_9-0-gts4lv.img is built with feature updates for Android 9 based devices compiled for TWRP and up to date support for Android 11.)

Considering incompatible TWRP and your issue / fix, I suspect now that boot.img may have been corrupted...

As I mentioned, zip installer Magisk patch won't touch recovery partition, but it will install Magisk binary in a basic ('manually added') ramdisk in boot.img when OEM ramdisk is missing. (Magisk App will still report Ramdisk, No to indicate no OEM ramdisk in boot partition.) Also, if you earlier patched AP without selecting Recovery mode, the same redundant Magisk/ramdisk addition will be made to boot.img.

I think it's likely that this redundant code in boot partition is somehow causing your issue. Boot contains kernel as well as ramdisk on all devices other than your special case non-A/B legacy SAR, and if Magisk patching was applied or Magisk zip installer was run, you now have both too. Significantly, the hardware blobs (for system handshaking with hardware) are in the kernel.

It may be that incompatible TWRP has moved blobs / ramdisk or location addresses around in boot.img, or there are other unexpected affects... The fact that sideloaded Magisk zip installation fixes the issue is an indication that boot, and possibly kernel corruption is the problem source... If so, and things are working with fixed patched boot.img even though the Magisk/ramdisk patch is redundant*, I imagine that simply flashing the unpatched LOS boot.img to remove redundant patching altogether, should also have no further effect.

*Your device can't use it simply because bootloader is incompatible with ramdisk in boot, but A-only legacy SAR Xiaomi/other devices can use both Magisk in recovery and Magisk in boot as boot ramdisk compatibility has inexplicably been left or added to bootloader.

It is also possible (more remotely?) that something a properly flashed magisk zip installer sets up is actually required, but I think the corrupted redundant boot install theory may be right...

The other issue you linked also makes me suspicious of a similar cause as the user mentions patching boot.img(!?)... but it does further cloud the issue as user reports only Magisk 24+ affected... Moreover, he had issue w/ Android 11...  🤔

It would be interesting to know (just to crush the issue) if another dirty ROM flash (should overwrite all ADB sideloaded zip changes) excluding recovery has any effect or if issue returns. I'm guessing it won't, but I'm no seer...

Also, after a dirty ROM flash inc. LOS recovery, does the official TWRP method work without issues with LOS recovery retained (ie. using Magisk-patched LOS recovery image instead of TWRP)?

I read, but I'm no dev, and hopeless at interpreting logs etc... hope a dev can do that for you...

refs:
- https://forum.xda-developers.com/t/magisk-general-support-discussion.3432382/post-86606967
- https://topjohnwu.github.io/Magisk/boot.html
- https://topjohnwu.github.io/Magisk/details.html#magisk-booting-process

**Comment by Deewiant** [comment]
Thanks for helping clarify things! Those Magisk docs you linked were also great reads. And your forum post mentioning dtb patching at least gives me some idea of how Magisk could affect audio, since if the dtb is incorrectly patched then the kernel definitely won't find all devices as it should — it does still seem much more likely that the device wouldn't boot at all, but it's always possible that I simply got "lucky" in this respect.

---

A new OTA update arrived today. I let the updater install it, figuring it more or less amounts to the "dirty ROM flash" you were interested in. It seemed to go fine, but Magisk had trouble: I saw these messages show up in TWRP's display just before it rebooted:

https://github.com/topjohnwu/Magisk/blob/b3f05fd925a7b62a0bd02e132dd8b423b538d37e/scripts/addon.d.sh#L36-L40

Perhaps this is due to the Android 12 incompatibility you mentioned. I've noticed that TWRP is unable to decrypt /sdcard at least, which feels related.

So at this point Magisk says "Installed N/A" and as such I don't have root any more. But at least audio still works (matching my previous experiences where "N/A" means audio works and anything else means it doesn't, until I sideloaded Magisk and somehow got both working). Booting into recovery just brings me to TWRP.

What exactly would you want me to test at this point? Have Magisk patch the LOS recovery, reboot into TWRP to flash it over the recovery partition, and see what happens? Or something else?

**Comment by pndwal** [comment]
It occurred to me to ask, Can I assume you have only ever booted successfully to system with Magisk via timed key combo for recovery (now requiring usb connection) on your device?... Or have you ever got system with root from device powered off using normal power press?

Assuming you always needed to boot via recovery, Try patching LOS recovery from Magisk app (Recovery mode set) and flashing that (via Odin?)... I don't think you can patch AP for recovery as seems it doesn't contain recovery.img, unless you can unpack, add LOS recovery and recompress/pack... Anyway, might see if you can get root w/o audio issue...

(unpatched, so we can assume all images are 'stock'), then 


method work without issues with LOS recovery retained (ie. using Magisk-patched LOS recovery image instead of TWRP)?

**Comment by Deewiant** [comment]
As far as I recall getting root has always required going via recovery. Generally I've done this by booting directly to recovery from LOS or with `adb reboot recovery`, since the timed key combo is a hassle. In other words my usual OTA update procedure has been to let the updater do its job, which boots me to Android directly, and then immediately reboot to recovery afterwards to ensure I have root.

I grabbed `lineage-19.1-20220528-recovery-gts4lv.img`, patched it with Magisk ("Recovery mode set" is a bit of an unnecessary point since it's the only option Magisk gives on this device...), flashed that onto the recovery partition with TWRP, issued `adb reboot recovery` and now I once again have root together with working audio!

It makes me think that it might be related to TWRP not fully supporting Android 12, but given the #5389 report it doesn't seem as straightforward as that...

**Comment by pndwal** [comment]
Seems TWRP issue... But to determine ADB zip install (patched boot, more?) is not still actively helping, seems need this test:
> It would be interesting to know (just to crush the issue) if another dirty ROM flash (should overwrite all ADB sideloaded zip changes) excluding recovery has any effect or if issue returns.

After flash, delete everything from /data/adb and reboot (to force daemon to rebuild Magisk configuration) also...

**Comment by Deewiant** [comment]
Alright, I reflashed the image I was already on. That triggered this message from Magisk again: https://github.com/topjohnwu/Magisk/blob/b3f05fd925a7b62a0bd02e132dd8b423b538d37e/scripts/addon.d.sh#L36-L40

Then, `/data/adb` was encrypted (seems Lineage Recovery can't decrypt things fully either) but I wiped it all anyway. On boot (through "recovery"), audio still worked and Magisk gave this message: https://github.com/topjohnwu/Magisk/blob/b3f05fd925a7b62a0bd02e132dd8b423b538d37e/app/src/main/res/values/strings.xml#L219

I let it do that, and on reboot both audio and Magisk were fine.

Out of curiosity I re-flashed the same image yet again afterwards, which didn't seem to change anything: I got the same message from `addon.d.sh` again, and audio and Magisk still both work.

**Comment by pndwal** [comment]
So boot.img is now unpatched (as should be) and config structure incl. Magisk database, magisk.db, is rebuild and no issues!...

Seems incompatible TWRP for Android 11- is confirmed as your issue... 😝 Glad this issue is sorted.

As I've said, addon.d is supported only with zip installer Magisk installation, but neither zip installation nor addon.d can work w/ Magisk in recovery partition anyway, regardless of encryption...

> "Recovery mode set" is a bit of an unnecessary point since it's the only option Magisk gives on this device...
Do you mean box cannot be unchecked?

**Comment by Deewiant** [comment]
> Seems incompatible TWRP (support only for Android 11-) is confirmed as your issue... :stuck_out_tongue_closed_eyes: Glad this issue is sorted.

I suppose so, so I'll close this issue. #5389 still makes me suspicious there's something more to this, since that was on Android 11. But as I only have problems with an unsupported configuration and not with a supported configuration, I don't really have a right to complain here...

> As I've said, addon.d is supported only with zip installer Magisk installation, but neither zip installation nor addon.d can work w/ Magisk in recovery partition anyway, regardless of encryption...

Yeah, I just thought it might already know it's not going to work and thus not even try and not complain. But that's fine.

> > "Recovery mode set" is a bit of an unnecessary point since it's the only option Magisk gives on this device...
> 
> Did you mean box cannot be unchecked?

Apparently it can and I simply never realized. 🤦🏻

Thanks again for all the help!
